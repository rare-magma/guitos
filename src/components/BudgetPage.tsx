import { useEffect, useState } from "react";
import { Budget } from "./Budget";
import { useNavigate, useParams } from "react-router-dom";
import { Container, Row, Col, Alert } from "react-bootstrap";
import TableCard from "./TableCard";
import { Stat } from "./Stat";
import StatCard from "./StatCard";
import {
  calcAvailable,
  calcSaved,
  calcWithGoal,
  convertCsvToBudget,
  createNewBudget,
} from "../utils";
import { Income } from "./Income";
import { Expense } from "./Expense";
import NavBar from "./NavBar";
import Papa, { ParseError } from "papaparse";
import localforage from "localforage";
import { Option } from "react-bootstrap-typeahead/types/types";

function BudgetPage() {
  const [saved, setSaved] = useState(false);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [budget, setBudget] = useState<Budget | null>(null);

  const [budgetList, setBudgetList] = useState<Budget[]>([]);

  let budgetNameList: { id: string; name: string }[] = [];
  if (budgetList.length >= 1 && Array.isArray(budgetList)) {
    budgetNameList = budgetList.map((b: Budget) => {
      return { id: b.id, name: b.name };
    });
  }

  const [csvError, setCsvError] = useState<ParseError[] | undefined>([]);
  const [csvErrorFiles, setCsvErrorFiles] = useState<string[] | undefined>([]);
  const [show, setShow] = useState(false);
  const params = useParams();
  const name = String(params.name);
  const navigate = useNavigate();

  const handleIncomeChange = (item: Income) => {
    let newBudget: Budget;
    if (budget !== null) {
      newBudget = budget;
      newBudget.incomes = item;
      newBudget.stats.available = calcAvailable(newBudget);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        incomes: {
          ...budget.incomes,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  };

  const handleExpenseChange = (item: Expense) => {
    let newBudget: Budget;
    if (budget !== null) {
      newBudget = budget;
      newBudget.expenses = item;
      newBudget.stats.available = calcAvailable(newBudget);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        expenses: {
          ...budget.expenses,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  };

  const handleStatChange = (item: Stat) => {
    let newBudget: Budget;
    if (budget !== null) {
      newBudget = budget;
      newBudget.stats = item;
      newBudget.stats.available = calcAvailable(newBudget);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
          goal: item.goal,
          reserves: item.reserves,
        },
      });
    }
  };

  const handleRename = (newName?: string | null) => {
    let newBudget: Budget;
    if (budget !== null && newName) {
      newBudget = budget;
      newBudget.name = newName;
      setBudget({
        ...budget,
        name: newBudget.name,
      });
      save(budget);
    }
  };

  const handleNew = () => {
    const newBudget = createNewBudget();
    let newBudgetList: Budget[] = [];
    if (budgetList !== null) {
      newBudgetList = budgetList.concat(newBudget);
    } else {
      newBudgetList = newBudgetList.concat(newBudget);
    }
    localforage
      .setItem(newBudget.id, newBudget)
      .then(() => {
        setBudget(newBudget);
        setBudgetList(newBudgetList);
        navigate("/" + newBudget.name);
        navigate(0);
      })
      .catch((e) => setError(e.message));
  };

  const handleClone = () => {
    if (budget !== null) {
      const newBudget = budget;
      newBudget.id = crypto.randomUUID();
      newBudget.name = budget.name + "-clone";
      let newBudgetList: Budget[] = [];
      if (budgetList !== null) {
        newBudgetList = budgetList.concat(newBudget);
      } else {
        newBudgetList = newBudgetList.concat(newBudget);
      }
      localforage
        .setItem(newBudget.id, newBudget)
        .then(() => {
          setBudget(newBudget);
          setBudgetList(newBudgetList);
          navigate("/" + newBudget.name);
          navigate(0);
        })
        .catch((e) => setError(e.message));
    }
  };

  const handleRemove = (toBeDeleted: string) => {
    localforage
      .removeItem(toBeDeleted)
      .then(() => {
        const newBudgetList = budgetList
          .filter((item: Budget) => item.id !== toBeDeleted)
          .sort((a, b) => a.name.localeCompare(b.name))
          .reverse();

        setBudgetList(newBudgetList);
        setBudget(newBudgetList[0]);
        if (newBudgetList.length >= 1) {
          navigate("/" + newBudgetList[0].name);
          navigate(0);
        }
      })
      .catch((e) => {
        console.error(e);
        setError(e.message);
      });
  };

  const handleSelect = (budget: Option[]) => {
    const selectedBudget = budget as unknown as Budget[];
    const filteredList = budgetList.filter(
      (item: Budget) => item.id === selectedBudget[0].id
    );
    console.log(JSON.stringify(filteredList));
    console.log(JSON.stringify(selectedBudget));
    setBudget(filteredList[0]);
    navigate("/" + selectedBudget[0].name);
    navigate(0);
  };

  const handleUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    const importedFiles = e.target.files;
    const newBudgetList: string[] = [];
    if (importedFiles === null) {
      return;
    }

    for (let i = 0; i < importedFiles.length; i++) {
      let newBudget: Budget;
      const file = importedFiles[i];
      const reader = new FileReader();
      reader.readAsText(file, "UTF-8");

      reader.onloadend = () => {
        if (reader.result !== null) {
          if (file.type === "text/csv") {
            const csvObject = Papa.parse(reader.result as string, {
              header: true,
              skipEmptyLines: "greedy",
            });
            if (csvObject.errors.length > 0) {
              setCsvError(csvError?.concat(csvObject.errors));
              setCsvErrorFiles(csvErrorFiles?.concat(file.name));
              console.error("error importing " + file.name);
              for (const i in csvObject.errors) {
                const errormsg = csvObject.errors[i] as unknown as ParseError;
                console.error(
                  "on row " + errormsg.row + ": " + errormsg.message
                );
              }
              setShow(true);
            }

            newBudget = convertCsvToBudget(
              csvObject.data as string[],
              file.name.slice(0, -4)
            );
            newBudgetList.push(newBudget as unknown as string);
            localforage
              .setItem(newBudget.id, newBudget)
              .then(() => {
                setBudgetList(newBudgetList as unknown as Budget[]);
                setBudget(budgetList[-1]);
              })
              .catch((e) => setError(e.message));
          } else {
            const list = JSON.parse(reader.result as string);
            list.forEach((b: string) => {
              localforage
                .setItem((b as unknown as Budget).id, b)
                .then(() => {
                  setBudgetList(budgetList.concat(newBudget));
                  setBudget(budgetList[0]);
                })
                .catch((e) => setError(e.message));
            });
          }
        }
      };
    }
  };

  const handleDownload = () => {
    let filename = new Date().toISOString();
    filename = `guitos-${filename}.json`;
    const url = window.URL.createObjectURL(
      new Blob([JSON.stringify(budgetList)])
    );
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", filename);
    document.body.appendChild(link);
    link.click();
  };

  const save = (budget: Budget) => {
    const newBudgetList = budgetList.map((data: Budget) => {
      if (budget.id === data.id) {
        return budget;
      } else {
        return data;
      }
    });
    localforage
      .setItem(budget.id, budget)
      .then(() => {
        setBudgetList(newBudgetList);
        setSaved(true);
      })
      .catch((e) => setError(e.message));
  };

  useEffect(() => {
    setLoading(true);
    try {
      if (budgetList.length >= 1 && Array.isArray(budgetList)) {
        budgetList
          .filter((budget: Budget) => budget.name === name)
          .map((data: Budget) => {
            setBudget(data);
            save(data);
          });
        setLoading(false);
      } else {
        let list: Budget[] = [];
        localforage
          .iterate((value) => {
            list = list.concat(value as unknown as Budget);
          })
          .then(() => {
            setBudgetList(list);
            if (list.length >= 1 && Array.isArray(list)) {
              list
                .filter((budget: Budget) => budget.name === name)
                .map((data: Budget) => {
                  setBudget(data);
                });
            }
            setLoading(false);
          })
          .catch((e) => console.error(e.message));
      }
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (e: any) {
      console.error(e);
      setError(e);
      setLoading(false);
    }
  }, [name, budget]);

  return (
    <div>
      <>
        <NavBar
          selected={budget?.name || undefined}
          id={budget?.id || undefined}
          budgetNameList={budgetNameList}
          onRename={(e) => {
            handleRename(e);
          }}
          onClone={() => {
            handleClone();
          }}
          onDownload={() => {
            handleDownload();
          }}
          onNew={() => {
            handleNew();
          }}
          onUpload={(e) => {
            handleUpload(e);
          }}
          onRemove={(e) => {
            handleRemove(e);
          }}
          onSelect={(e) => {
            handleSelect(e);
          }}
        />

        {loading && (
          <div className="center-page">
            <span className="spinner primary"></span>
            <p>Select an item from the list...</p>
          </div>
        )}

        {error && (
          <Alert variant="danger" onClose={() => setShow(false)} dismissible>
            <Alert.Heading>Error</Alert.Heading>
          </Alert>
        )}

        {/* {csvError &&
          show &&
          csvErrorFiles &&
          csvErrorFiles.map((file) => (
            {csvError.map((error) => (
                <Alert
                  variant="danger"
                  onClose={() => setShow(false)}
                  dismissible
                >
                  <Alert.Heading>Error while importing {file}:</Alert.Heading>
                  <p>{error.row}</p>;<p>{error.message}</p>;
                </Alert>
              ))
            }

          ))} */}

        {budget && (
          <Container>
            <div className="mt-3" />
            <Row>
              <Col md="6">
                <div className="card-columns">
                  <StatCard stat={budget.stats} onChange={handleStatChange} />

                  <div className="mt-3" />

                  <TableCard
                    items={budget.incomes}
                    header="Revenue"
                    onChange={handleIncomeChange}
                  />
                  <div className="mt-3" />
                </div>
              </Col>

              <Col md="6">
                <TableCard
                  items={budget.expenses}
                  header="Expenses"
                  onChange={handleExpenseChange}
                />
              </Col>
            </Row>
          </Container>
        )}
      </>
    </div>
  );
}

export default BudgetPage;
