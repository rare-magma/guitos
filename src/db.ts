import localforage from "localforage";

export const budgetsDB = localforage.createInstance({
  name: "guitos",
  storeName: "budgets",
});

export const optionsDB = localforage.createInstance({
  name: "guitos",
  storeName: "options",
});
