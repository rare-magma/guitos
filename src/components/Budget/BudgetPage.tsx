import Papa from "papaparse";
import { Suspense, lazy, useEffect, useRef, useState } from "react";
import { Col, Container, Row, ToastContainer } from "react-bootstrap";
import { useHotkeys } from "react-hotkeys-hook";
import { useParams } from "react-router-dom";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { budgetsDB, optionsDB } from "../../context/db";
import {
  budgetToCsv,
  calcAutoGoal,
  calcAvailable,
  calcSaved,
  calcWithGoal,
  convertCsvToBudget,
  createBudgetNameList,
  createNewBudget,
  roundBig,
  userLang,
} from "../../utils";
import ErrorModal, { CsvError, JsonError } from "../ErrorModal/ErrorModal";
import LandingPage from "../LandingPage/LandingPage";
import Loading from "../Loading/Loading";
import NavBar, { SearchOption } from "../NavBar/NavBar";
import Notification from "../Notification/Notification";
import { Stat } from "../StatCard/Stat";
import StatCard from "../StatCard/StatCard";
import { Expense } from "../TableCard/Expense";
import { Income } from "../TableCard/Income";
import TableCard from "../TableCard/TableCard";
import { Budget } from "./Budget";
// import { useWhatChanged } from "@simbathesailor/use-what-changed";

const ChartsPage = lazy(() => import("../ChartsPage/ChartsPage"));

function BudgetPage() {
  const inputRef = useRef<HTMLInputElement>(null);

  const [loading, setLoading] = useState(true);
  const [showGraphs, setShowGraphs] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [csvError, setCsvError] = useState<CsvError[]>([]);
  const [jsonError, setJsonError] = useState<JsonError[]>([]);
  const [show, setShow] = useState(false);
  const [focus, setFocus] = useState("");

  const [notifications, setNotifications] = useState<
    {
      show: boolean;
      id?: string;
      body?: string;
    }[]
  >([]);

  const { budget, setBudget, budgetList, setBudgetList, setBudgetNameList } =
    useBudget();
  const params = useParams();
  const name = String(params.name);

  const showCards = !loading && !showGraphs && budget?.id;

  const { setIntlConfig, handleCurrency } = useConfig();

  useHotkeys("escape", (e) => !e.repeat && setNotifications([]), {
    preventDefault: true,
  });
  useHotkeys("s", (e) => !e.repeat && handleExportJSON(), {
    preventDefault: true,
  });
  useHotkeys("d", (e) => !e.repeat && handleExportCSV(), {
    preventDefault: true,
  });
  useHotkeys("a", (e) => !e.repeat && !showGraphs && handleNew(), {
    preventDefault: true,
  });
  useHotkeys("c", (e) => !e.repeat && !showGraphs && handleClone(), {
    preventDefault: true,
  });
  useHotkeys(
    "i",
    (e) => !e.repeat && !showGraphs && budget && setShowGraphs(true),
    {
      preventDefault: true,
    },
  );

  function handleError(e: unknown) {
    if (e instanceof Error) setError(e.message);
    setShow(true);
  }

  function handleIncomeChange(item: Income) {
    let newBudget: Budget | undefined;
    if (budget) {
      newBudget = budget;
      newBudget.incomes = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        incomes: {
          ...budget.incomes,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  }

  function handleExpenseChange(item: Expense) {
    let newBudget: Budget;
    if (budget) {
      newBudget = budget;
      newBudget.expenses = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        expenses: {
          ...budget.expenses,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  }

  function handleStatChange(item: Stat | undefined) {
    let newBudget: Budget;
    if (budget && item) {
      newBudget = budget;
      newBudget.stats = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
          goal: item.goal,
          reserves: item.reserves,
        },
      });
    }
  }

  function handleAutoGoal(item: Stat | undefined) {
    let newBudget: Budget;
    if (budget && item) {
      newBudget = budget;
      newBudget.stats = item;
      newBudget.stats.goal = calcAutoGoal(budget);
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
          goal: newBudget.stats.goal,
          reserves: newBudget.stats.reserves,
        },
      });
    }
  }

  function handleRename(newName?: string | null) {
    let newBudget: Budget;
    if (budget && newName) {
      newBudget = {
        ...budget,
        name: newName,
      };
      setBudget(newBudget);
    }
  }

  function handleNew() {
    const newBudget = createNewBudget();

    let newBudgetList: Budget[] = [];
    if (budgetList) {
      newBudgetList = budgetList.concat(newBudget);
    } else {
      newBudgetList = newBudgetList.concat(newBudget);
    }

    setBudget(newBudget);
    setBudgetList(newBudgetList);
    setBudgetNameList(createBudgetNameList(newBudgetList));

    setNotifications([
      ...notifications,
      {
        show: true,
        id: crypto.randomUUID(),
        body: `created "${newBudget.name}" budget`,
      },
    ]);
  }

  function handleClone() {
    if (budget) {
      const newBudget = {
        ...budget,
        id: crypto.randomUUID(),
        name: budget.name + "-clone",
      };

      let newBudgetList: Budget[] = [];
      if (budgetList) {
        newBudgetList = budgetList.concat(newBudget);
      } else {
        newBudgetList = newBudgetList.concat(newBudget);
      }

      setNotifications([
        ...notifications,
        {
          body: `cloned "${budget.name}" budget`,
          id: crypto.randomUUID(),
          show: true,
        },
      ]);
      setBudget(newBudget);
      setBudgetList(newBudgetList);
      setBudgetNameList(createBudgetNameList(newBudgetList));
    }
  }

  function handleRemove(toBeDeleted: string) {
    budgetList &&
      budgetsDB
        .removeItem(toBeDeleted)
        .then(() => {
          const newBudgetList = budgetList
            .filter((item: Budget) => item.id !== toBeDeleted)
            .sort((a, b) => a.name.localeCompare(b.name))
            .reverse();

          setBudgetList(newBudgetList);
          setBudgetNameList(
            createBudgetNameList(newBudgetList as unknown as Budget[]),
          );

          setNotifications([
            ...notifications,
            {
              body: `deleted "${budget?.name}" budget`,
              id: crypto.randomUUID(),
              show: true,
            },
          ]);

          if (newBudgetList.length >= 1) {
            setBudget(newBudgetList[0]);
          } else {
            setBudget(undefined);
          }
        })
        .catch((e: unknown) => {
          handleError(e);
        });
  }

  function handleSelect(selectedBudget: SearchOption[] | undefined) {
    if (selectedBudget && budgetList) {
      const filteredList = budgetList.filter(
        (item: Budget) => item.id === selectedBudget[0].id,
      );

      filteredList && setBudget(filteredList[0]);

      if (selectedBudget[0].item && selectedBudget[0].item.length > 0) {
        setFocus(selectedBudget[0].item);
      }
    }
  }

  useEffect(() => {
    const element = document.querySelector(`input[value="${focus}"]`);
    if (element !== null) {
      (element as HTMLElement).focus();
    }
  }, [focus]);

  function handleGo(step: number, limit: number) {
    const sortedList = budgetList?.sort((a, b) => a.name.localeCompare(b.name));
    if (budget) {
      const index = sortedList?.findIndex((b) => b.name.includes(budget.name));
      if (index !== limit && sortedList) {
        handleSelect([
          sortedList[(index ?? 0) + step] as unknown as SearchOption,
        ]);
      }
    }
  }

  function handleGoHome() {
    if (budget) {
      const name = new Date().toISOString();
      const index = budgetList?.findIndex((b) =>
        b.name.includes(name.slice(0, 7)),
      );
      const isSelectable = index !== -1 && budgetList && index;

      if (isSelectable) {
        handleSelect([budgetList[index] as unknown as SearchOption]);
      }
    }
  }

  function handleGoBack() {
    budgetList && handleGo(-1, 0);
  }

  function handleGoForward() {
    budgetList && handleGo(1, budgetList.length - 1);
  }

  function handleImportCsv(fileReader: FileReader, file: File) {
    const newBudgetList: Budget[] = [];
    const csvObject = Papa.parse(fileReader.result as string, {
      header: true,
      skipEmptyLines: "greedy",
    });

    const hasErrors = csvObject.errors.length > 0;

    if (hasErrors) {
      csvError.push({
        errors: csvObject.errors,
        file: file.name,
      });
      setCsvError(csvError);
      setShow(true);
      setLoading(false);

      return;
    }

    const newBudget = convertCsvToBudget(
      csvObject.data as string[],
      file.name.slice(0, -4),
    );
    newBudgetList.push(newBudget);
    save(newBudget);
    setBudgetList(newBudgetList);
    setBudgetNameList(createBudgetNameList(newBudgetList));
  }

  function handleImportJSON(fileReader: FileReader, file: File) {
    const newBudgetList: Budget[] = [];
    try {
      const list = JSON.parse(fileReader.result as string) as Budget[];
      list.forEach((b: Budget) => {
        newBudgetList.push(b);
        save(b);
      });
      setBudgetList(newBudgetList);
      setBudgetNameList(createBudgetNameList(newBudgetList));
    } catch (e) {
      setJsonError([{ errors: (e as string).toString(), file: file.name }]);
      setShow(true);
      setLoading(false);
    }
  }

  function handleImport(e: React.ChangeEvent<HTMLInputElement>) {
    setLoading(true);
    const importedFiles = e.target.files;
    if (importedFiles === null) {
      return;
    }
    for (const file of importedFiles) {
      const reader = new FileReader();
      reader.readAsText(file, "UTF-8");
      reader.onloadend = () => {
        if (reader.result !== null) {
          if (file.type === "text/csv") {
            handleImportCsv(reader, file);
          } else {
            handleImportJSON(reader, file);
          }
        }
      };
    }
  }

  function handleExport(t: string) {
    if (t === "csv") handleExportCSV();
    if (t === "json") handleExportJSON();
  }

  function handleExportJSON() {
    if (budget) {
      let filename = new Date().toISOString();
      filename = `guitos-${filename.slice(0, -5)}.json`;
      const url = window.URL.createObjectURL(
        new Blob([JSON.stringify(budgetList)]),
      );
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  function handleExportCSV() {
    if (budget) {
      const filename = `${budget.name}.csv`;

      const url = window.URL.createObjectURL(new Blob([budgetToCsv(budget)]));
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  function save(budget: Budget) {
    let list: Budget[] = [];
    budgetsDB
      .setItem(budget.id, budget)
      .then(() => {
        budgetsDB
          .iterate((value) => {
            list = list.concat(value as Budget);
          })
          .then(() => {
            setBudgetList(list);
            setBudgetNameList(createBudgetNameList(list));
          })
          .catch((e: unknown) => {
            handleError(e);
          });
      })
      .catch((e: unknown) => {
        handleError(e);
      });
  }

  function loadFromDb() {
    let list: Budget[] = [];

    budgetsDB
      .iterate((value) => {
        list = list.concat(value as Budget);
      })
      .then(() => {
        setBudgetList(list);
        setBudgetNameList(createBudgetNameList(list));

        if (name.trim() !== "undefined") {
          setBudget(list.filter((b: Budget) => b && b.name === name)[0]);
        } else {
          setBudget(
            list
              .sort((a, b) => a.name.localeCompare(b.name))
              .reverse()
              .filter((b: Budget) => b && b.id === list[0].id)[0],
          );
        }

        loadCurrencyOption();
        setLoading(false);
      })
      .catch((e) => {
        handleError(e);
      });
  }

  function loadBudget(list: Budget[]) {
    list.forEach((data: Budget) => {
      budgetsDB
        .getItem(data.id)
        .then((b) => setBudget(b as Budget))
        .catch((e) => {
          handleError(e);
        });
    });
  }

  function loadCurrencyOption() {
    optionsDB
      .getItem("currencyCode")
      .then((c) => {
        if (c) {
          handleCurrency(c as string);
          setIntlConfig({ locale: userLang, currency: c as string });
        }
      })
      .catch((e) => {
        handleError(e);
      });
  }

  // useWhatChanged([budget, name]);

  useEffect(() => {
    budget && save(budget);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [budget]);

  useEffect(() => {
    try {
      const shouldLoadBudgetsFromList =
        budgetList && budgetList.length >= 1 && Array.isArray(budgetList);

      if (shouldLoadBudgetsFromList) {
        if (name.trim() !== "undefined") {
          loadBudget(budgetList.filter((b: Budget) => b && b.name === name));
        } else {
          loadBudget(budgetList.slice(0));
        }

        loadCurrencyOption();
        setBudgetNameList(createBudgetNameList(budgetList));
        setLoading(false);
      } else {
        loadFromDb();
      }
    } catch (e: unknown) {
      handleError(e);
      setLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [name, loading]);

  return (
    <Container fluid style={{ zIndex: 1 }}>
      <ToastContainer
        className="p-2"
        position={"bottom-center"}
        style={{ zIndex: 100 }}
      >
        {notifications.map((notification) => {
          return (
            notification && (
              <Notification
                notification={notification}
                onShow={() => {
                  setNotifications(
                    notifications.filter((n) => n.id !== notification.id),
                  );
                }}
              />
            )
          );
        })}
      </ToastContainer>

      {!showGraphs && (
        <NavBar
          onRename={(e) => {
            handleRename(e);
          }}
          onClone={() => {
            handleClone();
          }}
          onExport={(t) => {
            handleExport(t);
          }}
          onGoBack={() => {
            handleGoBack();
          }}
          onGoHome={() => {
            handleGoHome();
          }}
          onGoForward={() => {
            handleGoForward();
          }}
          onNew={() => {
            handleNew();
          }}
          onImport={(e) => {
            handleImport(e);
          }}
          onRemove={(e) => {
            handleRemove(e);
          }}
          onSelect={(e) => {
            handleSelect(e);
          }}
        />
      )}

      <LandingPage
        loading={loading}
        inputRef={inputRef}
        onNew={handleNew}
        onImport={(e) => {
          handleImport(e);
        }}
      />

      <ErrorModal
        error={error}
        show={show}
        jsonError={jsonError}
        csvError={csvError}
        onShow={setShow}
        onError={() => {
          setJsonError([]);
          setCsvError([]);
        }}
      />

      {showGraphs && (
        <Suspense fallback={<Loading />}>
          <ChartsPage onShowGraphs={() => setShowGraphs(false)} />
        </Suspense>
      )}

      {showCards && (
        <Container key={budget.id}>
          <Row className="mt-1">
            <Col md="6">
              <div className="card-columns">
                <StatCard
                  key={budget?.expenses.total + budget?.incomes.total}
                  onChange={handleStatChange}
                  onAutoGoal={handleAutoGoal}
                  onShowGraphs={() => setShowGraphs(true)}
                />

                <div className="mt-3" />

                <TableCard
                  items={budget.incomes}
                  header="Revenue"
                  onChange={handleIncomeChange}
                />
                <div className="mt-3" />
              </div>
            </Col>

            <Col md="6" className="mb-3">
              <TableCard
                items={budget.expenses}
                header="Expenses"
                onChange={handleExpenseChange}
              />
            </Col>
          </Row>
        </Container>
      )}
    </Container>
  );
}

export default BudgetPage;
