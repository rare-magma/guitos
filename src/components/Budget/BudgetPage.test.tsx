import { cleanup, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { budgetsDB } from "../../db";
import {
  budgetContextSpy,
  redoMock,
  setBudgetMock,
  setNotificationsMock,
  testBudget,
  testBudgetClone,
  testBudgetContext,
  undoMock,
} from "../../setupTests";
import { BudgetPage } from "./BudgetPage";

describe("BudgetPage", () => {
  const comp = <BudgetPage />;

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", async () => {
    render(comp);
    const newButton = screen.getAllByRole("button", { name: "new budget" });
    await userEvent.click(newButton[0]);
    expect(screen.getByLabelText("delete budget")).toBeInTheDocument();
  });

  it("responds to new budget keyboard shortcut", async () => {
    render(comp);
    await waitFor(async () => {
      await userEvent.type(screen.getByTestId("header"), "a");
    });
    expect(setBudgetMock).toHaveBeenCalled();
  });

  it("removes budget when clicking on delete budget button", async () => {
    render(comp);
    await waitFor(async () => {
      await expect(budgetsDB.getItem(testBudget.id)).resolves.toEqual(
        testBudget,
      );
      const deleteButton = screen.getAllByRole("button", {
        name: "delete budget",
      });
      await userEvent.click(deleteButton[0]);
      await userEvent.click(
        screen.getByRole("button", { name: "confirm budget deletion" }),
      );
      await expect(budgetsDB.getItem(testBudget.id)).resolves.toBeNull();
    });
  });

  it("clones budget when clicking on clone budget button", async () => {
    render(comp);
    await waitFor(async () => {
      const newButton = screen.getAllByRole("button", { name: "new budget" });
      await userEvent.click(newButton[0]);

      const cloneButton = screen.getAllByRole("button", {
        name: "clone budget",
      });
      await userEvent.click(cloneButton[0]);
      expect(setBudgetMock).toHaveBeenCalledWith(testBudgetClone, true);
    });
  });

  it("responds to clone budget keyboard shortcut", async () => {
    render(comp);
    await waitFor(async () => {
      const newButton = screen.getAllByRole("button", { name: "new budget" });
      await userEvent.click(newButton[0]);

      await userEvent.type(screen.getByTestId("header"), "c");
      expect(setBudgetMock).toHaveBeenCalledWith(testBudgetClone, true);
    });
  });

  it("responds to undo change keyboard shortcut", async () => {
    cleanup();
    budgetContextSpy.mockReturnValue({ ...testBudgetContext, canUndo: true });
    render(comp);
    await waitFor(async () => {
      await userEvent.type(screen.getByTestId("header"), "u");
      expect(undoMock).toHaveBeenCalled();
    });
  });

  it("responds to redo change keyboard shortcut", async () => {
    cleanup();
    budgetContextSpy.mockReturnValue({ ...testBudgetContext, canRedo: true });
    render(comp);
    await waitFor(async () => {
      await userEvent.type(screen.getByTestId("header"), "r");
      expect(redoMock).toHaveBeenCalled();
    });
  });

  it("responds to clear notifications keyboard shortcut", async () => {
    render(comp);
    setNotificationsMock.mockClear();
    await waitFor(async () => {
      await userEvent.type(screen.getByTestId("header"), "{Escape}");
      expect(setNotificationsMock).toHaveBeenCalledWith([]);
    });
  });

  it("responds to show graphs keyboard shortcut", async () => {
    render(comp);
    await waitFor(async () => {
      await userEvent.type(screen.getByTestId("header"), "i");
      expect(screen.getByRole("status")).toBeInTheDocument();
    });
  });
});
