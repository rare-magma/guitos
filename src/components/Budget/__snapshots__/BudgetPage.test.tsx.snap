// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BudgetPage > matches snapshot 1`] = `<BudgetPage />`;
